spread/openstack_test.go

```diff
@@ -0,0 +1,153 @@
+package spread_test
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/go-goose/goose/v5/glance"
+
+	"github.com/snapcore/spread/spread"
+
+	. "gopkg.in/check.v1"
+)
+
+func newOpenstack() *spread.OpenstackProvider {
+	prj := &spread.Project{}
+	b := &spread.Backend{}
+	opts := &spread.Options{}
+
+	o := spread.Openstack(prj, b, opts)
+	return o.(*spread.OpenstackProvider)
+}
+
+type openstackSuite struct{}
+
+var _ = Suite(&openstackSuite{})
+
+func (s *openstackSuite) TestTrivial(c *C) {
+	o := newOpenstack()
+	c.Check(o, NotNil)
+}
+
+func (s *openstackSuite) TestOpenstackName(c *C) {
+	restore := spread.MockTimeNow(func() time.Time {
+		return time.Date(2007, 8, 22, 11, 59, 58, 987654321, time.UTC)
+	})
+	defer restore()
+
+	name := spread.OpenstackName()
+	c.Check(name, Equals, "aug221159-987654")
+}
+
+var opstErr1 = errors.New(`caused by: requesting token failed
+caused by: Resource at https://keystone.bos01.canonistack.canonical.com:5000/v3/tokens not found
+caused by: request (https://keystone.bos01.canonistack.canonical.com:5000/v3/tokens) returned unexpected status: 404; error info: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
+<title>404 Not Found</title>
+<h1>Not Found</h1>
+<p>The requested URL was not found on the server.  If you entered the URL manually please check your spelling and try again.</p>`)
+
+var opstErr2 = errors.New(`caused by: requesting token failed
+caused by: request (https://keystone.bos01.canonistack.canonical.com:5000/v3/tokens) returned unexpected status: 503; error info: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
+<html><head>
+<title>503 Service Unavailable</title>
+</head><body>
+<h1>Service Unavailable</h1>
+<p>The server is temporarily unable to service your
+request due to maintenance downtime or capacity
+problems. Please try again later.</p>
+<hr>
+<address>Apache/2.4.29 (Ubuntu) Server at 10.48.7.10 Port 5000</address>
+</body></html>`)
+
+func (s *openstackSuite) TestErrorMsg(c *C) {
+	msg1 := spread.ErrorMsg(opstErr1)
```

Comment:
Gustavo: only have full output when enabling debug output!

Reply:
well, we are already returning the first "caused by" line

Reply:
I updated the code a bit now to follow Gustavos advise. Fwiw, the existing code was already using the last "caused by" because the for() loop did not exiting on the first match so it was already picking the last :) I also added a comment that we need to see if how well the error parsing works in practice (only real world errors will tell). I also added that the full error is always show when debug is enabled.